source/dem/insertion.cc

```diff
@@ -197,5 +221,131 @@ Insertion<dim>::calculate_insertion_domain_maximum_particle_number(
+          // Loop through its vertices
+          for (unsigned int vertex_id = 0; vertex_id < cell->n_vertices();
+               ++vertex_id)
+            {
+              Point<3> vertex = point_nd_to_3d(cell->vertex(vertex_id));
+
+              // Check if the n_th vertex is in the box
+              if (p_min[0] <= vertex[0] && vertex[0] <= p_max[0] &&
+                  p_min[1] <= vertex[1] && vertex[1] <= p_max[1] &&
+                  p_min[2] <= vertex[2] && vertex[2] <= p_max[2])
+                {
+                  partially_inside = true;
+                }
+              else
+                {
+                  completely_inside = false;
+                }
+            }
+          // If the cell is completely inside the clearing box, both bool will
+          // be true. We need to add the cell iterator to the
+          // in_the_clearing_box container.
+
+          // If the cell is partially inside, at least one of the "if" will have
+          // failed, thus the second bool will be false but the first one true.
+          // We need to add the cell iterator to the edge_of_box container.
+
+          // If the cell is outside, both bool will be false, and we do nothing.
+          if (completely_inside)
+            {
+              in_the_clearing_box.insert(cell);
+            }
+          else if (!completely_inside && partially_inside)
+            {
+              edge_of_box.insert(cell);
+            }
+        }
+    }
+}
+
+template <int dim>
+void
+Insertion<dim>::clear_particle_box(
+  Particles::ParticleHandler<dim> &particle_handler)
+{
+  std::vector<
+    typename dealii::Particles::ParticleHandler<dim>::particle_iterator>
+    to_remove_iterators;
+
+  // Reserve to the maximum number of particle on this proc
+  to_remove_iterators.reserve(particle_handler.n_locally_owned_particles());
+
+  // Loop over the first container
+  for (const auto &cell_in_box : in_the_clearing_box)
+    {
+      // Check if this cell has particles
+      auto particles_in_cell = particle_handler.particles_in_cell(cell_in_box);
+
+      if (!particles_in_cell.empty())
+        {
+          // Loop over the particle in the cell
+          for (auto particle_in_cell = particles_in_cell.begin();
+               particle_in_cell != particles_in_cell.end();
+               ++particle_in_cell)
+            {
+              // Since we know the cell is fully inside the box, we can
+              // delete every particles in it.
+
+              to_remove_iterators.push_back(particle_in_cell);
+            }
+        }
+    }
+
+  // Loop over the second container
+  for (auto cell_edge_of_box = edge_of_box.begin();
+       cell_edge_of_box != edge_of_box.end();
+       ++cell_edge_of_box)
+    {
+      // Check if this cell has particle
+      auto particles_in_cell =
+        particle_handler.particles_in_cell(*cell_edge_of_box);
+      const bool particles_exist_in_cell = !particles_in_cell.empty();
```

Comment:
Here you use a boolean, up there you use `!particles_in_cell.empty()`. Any specific reason?